tests/ContextTest.php

```diff
@@ -2,68 +2,152 @@
+        $this->assertAttributeEmpty('data', $context);
     }
 
-    public function testSetUserData()
+    public function testGetDataByPath()
+    {
+        $context = new Context(['foo' => ['bar' => 'baz']]);
+
+        $this->assertEquals('baz', $context->getDataByPath('[foo][bar]'));
+    }
+
+    public function testSetDataByPath()
     {
         $context = new Context();
+        $context->setDataByPath('[foo]', ['bar' => []]);
+
+        $this->assertAttributeEquals(['foo' => ['bar' => []]], 'data', $context);
 
-        $context->setUserData(['foo' => 'bar']);
-        $context->setUserData(['bar' => 'baz']);
+        $context->setDataByPath('[foo][bar]', 'baz');
 
-        $this->assertEquals(['bar' => 'baz'], $context->getUserData());
+        $this->assertAttributeEquals(['foo' => ['bar' => 'baz']], 'data', $context);
     }
 
-    public function testMergeExtraData()
+    public function testIsEmpty()
     {
         $context = new Context();
 
-        $context->mergeExtraData(['foo' => 'bar']);
-        $context->mergeExtraData(['baz' => 'bar']);
+        $this->assertTrue($context->isEmpty());
 
-        $this->assertEquals(['foo' => 'bar', 'baz' => 'bar'], $context->getExtraData());
+        $context->setData(['foo' => 'bar']);
+
+        $this->assertFalse($context->isEmpty());
+    }
+
+    public function testToArray()
+    {
+        $context = new Context(['foo' => 'bar']);
+
+        $this->assertEquals(['foo' => 'bar'], $context->toArray());
     }
 
-    public function testMergeExtraDataWithSameKey()
+    public function testJsonSerialize()
+    {
+        $context = new Context(['foo' => 'bar']);
+
+        $this->assertEquals('{"foo":"bar"}', json_encode($context));
+    }
+
+    public function testArrayLikeBehaviour()
     {
         $context = new Context();
 
-        $context->mergeExtraData(['foo' => 'bar']);
-        $context->mergeExtraData(['foo' => 'baz']);
+        $this->assertAttributeEquals([], 'data', $context);
+        $this->assertFalse(isset($context['foo']));
+
+        $throwedError = false;
+
+        // Accessing a key that does not exists in the data object should behave
+        // like accessing a non-existent key of an array
+        try {
+            $context['foo'];
+        } catch (\Exception $exception) {
+            if (class_exists(Notice::class)) {
+                $this->assertInstanceOf(Notice::class, $exception);
+            } else {
+                $this->assertInstanceOf(\PHPUnit_Framework_Error_Notice::class, $exception);
+            }
+
+            $this->assertStringStartsWith('Undefined index: foo', $exception->getMessage());
+
+            $throwedError = true;
+        }
+
+        $this->assertTrue($throwedError);
+
+        $context['foo'] = 'bar';
+
+        $this->assertAttributeEquals(['foo' => 'bar'], 'data', $context);
+        $this->assertTrue(isset($context['foo']));
+        $this->assertEquals('bar', $context['foo']);
+
+        unset($context['foo']);
```

Comment:
Is this needed?

Reply:
Yes it's needed to test that the `Context::offsetUnset` method works, but it was missing the assertion so nice catch!